### PR TITLE
Graph engine

### PR DESCRIPTION
**Pull request by leogrim**
No need to merge this now, this is to open it for comments.
@2is10 @gmethvin @andrewconner Look up TypeCode / TypeProvider, that pattern could be generalized if we implement Textual Ids.

@Yasu @stkem @yingjieMiao Any comment on the graph model?

### CODE REVIEW

File: kifi-backend/modules/graph/app/com/keepit/graph/model/TypeCode.scala [hunk 1]
```diff
@@ -0,0 +1,11 @@
...
+
+trait TypeProvider[+T] {
+  implicit def typeCode: TypeCode[T]
+  implicit val typeProvider: TypeProvider[T] = this
+  implicit def instanceToTypeProvider[I >: T](instance: I): TypeProvider[I] = this
+}
+
+case class TypeCode[+T](code: Symbol) {
+  override def toString() = code.name
+}
```
Comment (leogrim) [inline kifi-backend/modules/graph/app/com/keepit/graph/model/TypeCode.scala]: The pattern is similar to TypeTag's, except we can customize the TypeCode of each class.
By having TypeProvider extend a class' companion object, we can access the various implicits using context bounds (to make implicit values typeCode and typeProvider available) and view bounds (to have an instance of that class automatically respond to the typeCode method)

File: kifi-backend/modules/graph/app/com/keepit/graph/database/NeoEdge.scala [hunk 2]
```diff
@@ -0,0 +1,10 @@
+package com.keepit.graph.database
+
+import org.neo4j.graphdb.Relationship
+import com.keepit.graph.model.Edge
+
+case class NeoEdge[+S, +D, +E](relationship: Relationship) extends Edge[S, D, E] {
+  def source = NeoVertex[S](relationship.getStartNode())
+  def destination = NeoVertex[D](relationship.getEndNode())
+  def data = relationship.getProperty("data").asInstanceOf[E]
```
Comment (yingjieMiao) [inline kifi-backend/modules/graph/app/com/keepit/graph/database/NeoEdge.scala]: is this asInstanceOf method safe? Do we need a wrapper class to make sure this relationship contains data of type E?

Comment (leogrim) [inline kifi-backend/modules/graph/app/com/keepit/graph/database/NeoEdge.scala]: I make sure the type is correct at Edge and Vertex creations. But I'd be happy to have a better way, polymorphic serialization / deserialization is the main issue here.

File: kifi-backend/modules/graph/app/com/keepit/graph/database/NeoGraph.scala [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+    val node = graphDb.getNodeById(id.id)
+    NeoVertex[U](node)
+  }
+
+  def createEdges[S <: V, D <: V, F <: E <% TypeProvider[F]](edges: (VertexId[S], VertexId[D], F)*) = edges.map { case (source, destination, data) =>
+    val sourceNode = graphDb.getNodeById(source.id)
+    val destinationNode = graphDb.getNodeById(destination.id)
+    val relationshipType = DynamicRelationshipType.withName(data.typeCode.toString())
+    val relationship = sourceNode.createRelationshipTo(destinationNode, relationshipType)
+    NeoEdge[S, D, F](relationship)
```
Comment (yingjieMiao) [inline kifi-backend/modules/graph/app/com/keepit/graph/database/NeoGraph.scala]: Looks like in each graph[V, E], all vertices are of same type V (or inherited from V), say cat and dog But the NeoEdge definition gives me an impression that an edge can connect to vertexes of different types, say cat and car.

Comment (leogrim) [inline kifi-backend/modules/graph/app/com/keepit/graph/database/NeoGraph.scala]: Indeed Graph[V, E] means that all vertices hold data that inherits from V and all edges hold data that inherits from E. Indeed in this setup, the type of an edge is defined by its data, perhaps I should modify it to encompass the type of the vertices as well. Of have E be parametrized by S and D

Comment (yingjieMiao) [inline kifi-backend/modules/graph/app/com/keepit/graph/database/NeoGraph.scala]: That would be great. This way it's also consistent with the outgoingEdges method.  It wants to connect type U vertexes to type V vertexes. It doesn't force U and V to be inherited from same super class.

Comment (yingjieMiao) [inline kifi-backend/modules/graph/app/com/keepit/graph/database/NeoGraph.scala]: If E is parametrized as E[S, D]. Then this is asymmetric? E[S, D] not the same as E[D, S]? Looks like E and relationship are trying to describe the same thing. Maybe we can combine them somehow to kill one free variable? In createEdges, val relationshipType = DynamicRelationshipType.withName(data.typeCode.toString()), this is well defined. But the F seems a free variable, not tight to the edge data at all.  Maybe I missed something.

Comment (leogrim) [inline kifi-backend/modules/graph/app/com/keepit/graph/database/NeoGraph.scala]: Relationship is Neo4J's implementation of edges. I want to abstract it away from our own code.
F is a subtype of E, which is the superclass of all edges in the graph

**Comment by yingjieMiao** [comment]
Looks great overall!

**Comment by gmethvin** [comment]
It might be nice to have a few tests of this graph API, at least for the sake of seeing some examples of how it's used.

**Pull request closed without merging**